### DISCUSSION TITLE
403 Error on attempt to upload to public bucket.

### DISCUSSION THREAD

**Discussion by SouthCentralCode** [supabase/supabase#30228] [Questions]
Hi guys, i'm getting a 403 error

{
    "statusCode": "403",
    "error": "Unauthorized",
    "message": "new row violates row-level security policy"
}

From the code...  run from a client component on a public facing page on the web with no auth setup
 (nextjs 14)     "@supabase/supabase-js": "^2.45.4",

```
    const { data, error } = await client.storage
      .from('c_staging_bucket')
      .upload(`${fileUuid}/${file.name}`, file, {
        upsert: false,
        contentType: file.type,
      });
```

This is my policy
```
create policy "my_policy_name"
on storage.objects for insert to public with check (
    -- restrict bucket
    bucket_id = 'c_staging_bucket'
);

```

Is this the expected behaviour?

**Comment by SouthCentralCode** [comment]
as a note i've tried this on various new buckets.  

I also have no custom triggers setup that would cause knock on effect.

**Comment by GaryAustin1** [comment]
It is expected if you don't meet RLS to get that error.  Even public buckets need RLS for upload.  You have upsert false so do not need to meet select/update policies.

Is that your only insert policy.  That one does not show a TO clause (like authenticated/anon/public) so would default to public and work with anon or authenticated.

So.... are you going to the correct instance?  Can you see the request in the Storage or Gateway API logs?

**Comment by SouthCentralCode** [comment]
Hey, Thanks for the response.

That is the only RLS policy on that bucket. 

Where you mention I don't have "to clause"... 

but there is one..

_"on storage.objects for insert to public with check ("_

Here's the error log from the storage logs in the supabase backend in the expected instance.

`    
"error"
: [
      {
        
"message"
: 
"new row violates row-level security policy"
,
        
"name"
: 
"Error"
,
        
"raw"
: 
"{"metadata":{"query":"insert into \"objects\" (\"bucket_id\", \"metadata\", \"name\", \"owner\", \"owner_id\", \"user_metadata\", \"version\") values ($1, DEFAULT, $2, DEFAULT, DEFAULT, DEFAULT, $3)","code":"42501"},"code":"AccessDenied","httpStatusCode":403,"userStatusCode":400,"originalError":{"length":95,"name":"error","severity":"ERROR","code":"42501","file":"aclchk.c","line":"3650","routine":"aclcheck_error"},"error":"Unauthorized"}"
,
        
"stack"
: 
"Error: new row violates row-level security policy
    at Object.AccessDenied (/app/dist/internal/errors/codes.js:121:32)
    at DBError.fromDBError (/app/dist/storage/database/knex.js:542:37)
    at Function.<anonymous> (/app/dist/storage/database/knex.js:46:23)
    at Object.onceWrapper (node:events:634:26)
    at Function.emit (node:events:519:28)
    at Client_PG.<anonymous> (/app/node_modules/knex/lib/knex-builder/make-knex.js:304:10)
    at Client_PG.emit (node:events:531:35)
    at /app/node_modules/knex/lib/execution/internal/query-executioner.js:46:12
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Runner.ensureConnection (/app/node_modules/knex/lib/execution/runner.js:318:14)"

      }`

**Reply by GaryAustin1** [reply]
Sorry missed the to... Your RLS seems fine and pretty basic and not dependent on a user type.

Can you upload from the UI?
Can you try and upload with service_role key (and no user session) as a test?

It is saying RLS error, but I think if there is a grant issue on storage.objects it also can error with the same message.  Are you using any sort of custom tokens?  Are you using Prisma or Drizzle that might mess up the storage table grants?

**Comment by SouthCentralCode** [comment]
It turned out; that there was an error occurring in a public bucket policy that was causing a failure.
